### PR TITLE
Correctly count usage in recursive definitions

### PR DESCRIPTION
**Pull request by AliasQli**
This addresses issue #872. It also affects the "Go to References" function in VSCode.

This doesn't handle mutual recursion, otherwise usage counts would all become 0.

### CODE REVIEW

**Comment by AliasQli** [comment]
(But how do I write tests for this?)

File: ide/src/main/java/org/aya/ide/action/FindReferences.java [hunk 1]
```diff
@@ -31,6 +32,22 @@ public interface FindReferences {
...
+    var vars = Resolver.resolveVar(source, xy);
+    return findRefsOutsideDefs(vars.map(WithPos::data), libraries);
+  }
+
+  static @NotNull SeqView<SourcePos> findRefsOutsideDefs(
+    @NotNull SeqView<AnyVar> vars,
+    @NotNull SeqView<LibraryOwner> libraries
+  ) {
+    var defPos = vars.filterIsInstance(DefVar.class).map(def -> def.concrete.entireSourcePos());
+    return findRefs(vars, libraries).filter(ref -> !defPos.map(pos -> pos.containsIndex(ref)).contains(true));
```
Comment (AliasQli) [inline ide/src/main/java/org/aya/ide/action/FindReferences.java:48]: I wonder if this would do extra work because Java is not lazy.

Comment (imkiva) [reply ide/src/main/java/org/aya/ide/action/FindReferences.java:48]: What do you mean by "extra work"?

**Comment by imkiva** [comment]
> (But how do I write tests for this?)

We don't have a test framework for lsp features yet, see #433

**Comment by AliasQli** [comment]
In lazy languages like Haskell, evaluation would terminate the first time `pos.containsIndex(ref)` evaluates to `true`. Is it the same here?2023年5月30日 20:43，Kiva ***@***.***>写道：
@imkiva commented on this pull request.



In ide/src/main/java/org/aya/ide/action/FindReferences.java:
> @@ -31,6 +32,22 @@ public interface FindReferences {
     });
   }
 
+  static @NotNull SeqView<SourcePos> findRefsOutsideDefs(
+    @NotNull LibrarySource source,
+    @NotNull SeqView<LibraryOwner> libraries, XY xy
+  ) {
+    var vars = Resolver.resolveVar(source, xy);
+    return findRefsOutsideDefs(vars.map(WithPos::data), libraries);
+  }
+
+  static @NotNull SeqView<SourcePos> findRefsOutsideDefs(
+    @NotNull SeqView<AnyVar> vars,
+    @NotNull SeqView<LibraryOwner> libraries
+  ) {
+    var defPos = vars.filterIsInstance(DefVar.class).map(def -> def.concrete.entireSourcePos());
+    return findRefs(vars, libraries).filter(ref -> !defPos.map(pos -> pos.containsIndex(ref)).contains(true));

.filter(ref -> !defPos.map(pos -> pos.containsIndex(ref)).contains(true)) can be replaced by
.filter(ref -> defPos.noneMatch(pos -> pos.containsIndex(ref)))



In ide/src/main/java/org/aya/ide/action/FindReferences.java:
> @@ -31,6 +32,22 @@ public interface FindReferences {
     });
   }
 
+  static @NotNull SeqView<SourcePos> findRefsOutsideDefs(
+    @NotNull LibrarySource source,
+    @NotNull SeqView<LibraryOwner> libraries, XY xy
+  ) {
+    var vars = Resolver.resolveVar(source, xy);
+    return findRefsOutsideDefs(vars.map(WithPos::data), libraries);
+  }
+
+  static @NotNull SeqView<SourcePos> findRefsOutsideDefs(
+    @NotNull SeqView<AnyVar> vars,
+    @NotNull SeqView<LibraryOwner> libraries
+  ) {
+    var defPos = vars.filterIsInstance(DefVar.class).map(def -> def.concrete.entireSourcePos());
+    return findRefs(vars, libraries).filter(ref -> !defPos.map(pos -> pos.containsIndex(ref)).contains(true));

What do you mean by "extra work"?

**Comment by imkiva** [comment]
Yes. `SeqView` is lazy.

File: ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java [hunk 2]
```diff
@@ -328,7 +328,7 @@ public Optional<SignatureHelp> signatureHelp(TextDocumentPositionParams params)
     var source = find(params.textDocument.uri);
     if (source == null) return Optional.empty();
     return Optional.of(FindReferences
-      .findRefs(source, libraries.view(), LspRange.pos(params.position))
+      .findRefsOutsideDefs(source, libraries.view(), LspRange.pos(params.position))
```
Comment (ice1000) [inline ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java:331]: This should be like before, recursive usages are still usages

Comment (AliasQli) [reply ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java:331]: Then it would be strange: you clicked "1 usage" and found two usages

Comment (ice1000) [reply ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java:331]: Sounds acceptable to me

Comment (AliasQli) [reply ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java:331]: Though I still feel strange...
I'm in bed now. You can push to this brance to revert the change.

Comment (ice1000) [reply ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java:331]: I'll leave it to you ;)

**Comment by AliasQli** [comment]
What does this milestone mean?

**Comment by ice1000** [comment]
> What does this milestone mean?

**Comment by ice1000** [comment]
> What does this milestone mean?

We create releases on maven central from time to time. If you're not expecting to use Aya utilities in other projects, you can forget about this.

My apology for the misoperation

File: ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java (hunk 2 above)
Comment (imkiva) [reply ide-lsp/src/main/java/org/aya/lsp/server/AyaLanguageServer.java]: We may change the message to "%d other usages".

**Pull request merged**